File: get_ensembl.pl
#!/usr/bin/env perl

# A script to retrieve the current_mysql data for the latest ensembl release and push it to a local MySQL database
# Coded by Steve Moss
# gawbul@gmail.com
# 7th February 2011

# ToDo
#

###########################################################################
# * YOU NEED AT LEAST 2TB OF FREE SPACE FOR A SINGLE RELEASE OF ENSEMBL * #
###########################################################################

# make life easier
use warnings;
use strict;

# imports needed
use Net::FTP;
use Cwd;
use DBI;
use DBD::mysql;
use File::Spec;
use IO::Uncompress::Gunzip qw(gunzip $GunzipError);
use Time::HiRes qw(gettimeofday);

##############################################
# Variables here - change values as required #
##############################################

# variables for FTP connection
my $host = "ftp.ensembl.org";
my $username = "anonymous";
my $password = undef;
my $ftpdir = "/pub/current_mysql";

# variables for MySQL connection
my $sql_host = "localhost";
my $sql_port = 3306;
my $sql_username = "root";
# ask for MySQL server password instead of storing in plain text
print "What is your MySQL password (press enter for none)?: ";
my $sql_password = <STDIN>;
chop($sql_password);

# other variables
my $data_dir = "/Volumes/DATA/current_mysql"; ####### CHANGE THIS TO THE DIRECTORY YOU WANT TO STORE YOUR FILES IN #######

my $start_time = gettimeofday;

########################
# FTP stuff below here #
########################

# connect to ensembl ftp server
my $ftp = Net::FTP->new($host, KeepAlive=>1) or die "Error connecting to $host: $!";
 
# ftp login
$ftp->login($username, $password) or die "Login failed: $!";
 
# chdir to $ftpdir
$ftp->cwd($ftpdir) or die "Can't go to $ftpdir: $!";

# get list of ftp directories
my @dirs = $ftp->ls();
my %ftp_files = ();

# traverse the directories and just match core
foreach my $dir (@dirs) {
	# check if we have a core directory
	# this limits the downloaded data to the core databases
	# there is nothing stopping you removing this regex and downloading all the MySQL databases from EnsEMBL
	# obviously this depends on both the amount of time and space you have available
	#if ($dir =~ /[a-z]+_[a-z]+_core_[0-9]{2}_[0-9[a-z]{0,}/ || $dir =~ /ensembl_[a-z]+_[0-9]{2}/) { ####### CHANGE THIS REGEX IF YOU ONLY WANT SPECIFIC DATABASE #######
		# chdir to $dir
		my $newdir = $ftpdir . "/" . $dir;
		$ftp->cwd($newdir) or die "Can't go to $newdir: $!";

		print "Retrieving FTP directory structure for $dir...\n";

		# get file list
		my @files = $ftp->ls();
		
		# add array to hash
		$ftp_files{$dir} = \@files;
		
		# return to FTP root
		$ftp->cwd() or die "Can't go to FTP root: $!";
	#}	
}
print "Done!\n\n";

#-- close ftp connection
$ftp->quit or die "Error closing ftp connection: $!";

########################
# File retrieval stuff #
########################

# get directory list
@dirs = keys %ftp_files;
@dirs = sort {$a cmp $b} @dirs;

foreach my $dir (@dirs) {
	# build local directory structure
	unless (-d $data_dir) {
		mkdir $data_dir;
	}
	my $path = File::Spec->catfile($data_dir, $dir);
	unless (-d $path) {
		mkdir $path;
	}

	print "Retrieving files for $dir...\n";	

	my $files_ref = $ftp_files{$dir};
	my @files = @$files_ref;

	foreach my $file (@files) {
		unless (-e substr($file, 0, -3)) {
			# change to correct directory
			chdir $path;
			
			# retrieve the file
			system("wget -t 0 -c -N ftp://$host$ftpdir/$dir/$file");
		}
	}
}			
print "Done!\n\n";

##########################
# Extract the data files #
##########################

# change to data dir
chdir $data_dir;

# get directory listing
opendir(my $dh, $data_dir) or die "Can't opendir $data_dir: $!";
@dirs = grep {!/^\./ && -d "$data_dir/$_" } readdir($dh);
closedir $dh;

# sort directories
@dirs = sort {$a cmp $b} @dirs;

# traverse directories
foreach my $dir (@dirs) {
	# change into each directory in turn
	chdir "$data_dir/$dir";
	
	print "Extracting data for $dir (please be patient)...\n";

	# get list of files
	opendir($dh, "$data_dir/$dir") or die "Can't opendir $data_dir/$dir: $!";
	my @files = grep {!/^\./ && -f "$data_dir/$dir/$_" } readdir($dh);
	closedir $dh;
	@files = sort {$a cmp $b} @files;

	# populate tables in turn
	foreach my $file (@files) {
		if ($file =~ /\.gz$/) {
			my $input = $file;
			my $output = substr($file, 0, -3);
			my $status = 1;
			while (!(-e "$data_dir/$dir/$output")) {
				gunzip $input => $output or $status = 0;
				
				# if error in unzip then retrieve file again
				if ($status == 0) {
					# delete output file
					unlink $output;
					
					# get file again
					system("wget -t 2 -c -N ftp://$host/$ftpdir/$dir/$input");
					
					# update status
					$status = 1;
				}
			}
			unlink $input;
		}
	}
}
print "Done!\n\n";
chdir getcwd();

##########################
# MySQL stuff below here #
##########################

# change to data dir
chdir $data_dir;

# get directory listing
opendir($dh, $data_dir) or die "Can't opendir $data_dir: $!";
@dirs = grep {!/^\./ && -d "$data_dir/$_" } readdir($dh);
closedir $dh;

# sort directories
@dirs = sort {$a cmp $b} @dirs;

# traverse directories
foreach my $dir (@dirs) {
	# change into each directory in turn
	chdir "$data_dir/$dir";
	
	# create the database based on the dir name
	print "Creating database for $dir...\n";
	# setup database connection
	my $dsn = "DBI:mysql::$sql_host:$sql_port";
	my $dbh = DBI->connect($dsn, $sql_username, $sql_password) or die "Unable to connect: $DBI::errstr\n";
	$dbh->do("DROP DATABASE IF EXISTS $dir");
	$dbh->do("CREATE DATABASE IF NOT EXISTS $dir");
	$dbh->disconnect();
	
	# populate database with tables from .sql file
	print "Building database structure for $dir...\n";
	my $sql_file = File::Spec->catfile($data_dir, $dir, $dir . ".sql");
	system("mysql -h $sql_host -P $sql_port -u $sql_username --password=$sql_password $dir \< $sql_file");
			
	# get list of files
	opendir(my $dh, "$data_dir/$dir") or die "Can't opendir $data_dir/$dir: $!";
	my @files = grep {!/^\./ && -f "$data_dir/$dir/$_" } readdir($dh);
	closedir $dh;
	@files = sort {$a cmp $b} @files;
	
	# populate tables in turn
	foreach my $file (@files) {
		if ($file =~ /.*?\.txt$/) {
			# get variables
			my $sql_file = File::Spec->catfile($data_dir, $dir, $file);
			my $table = substr($file, 0, -4);
			
			# build query and execute
			print "Loading data in $sql_file into $dir...\n";
			system("mysqlimport -h $sql_host -P $sql_port -u $sql_username --password=$sql_password $dir $sql_file");
			#unlink $sql_file;
		}
	}
	
}
print "Done!\n\n";
chdir getcwd();

my $end_time = gettimeofday;
my $total_time = $end_time - $start_time;

print "Finished in $total_time seconds!\n";

File: get_ensembl.py
#!/usr/bin/env python

'''
get_ensembl.py
A script to retrieve the latest EnsEMBL release MySQL data from the EnsEMBL ftp server
Coded by Steve Moss
gawbul@gmail.com
20th April 2011
'''

# imports
from ftplib import FTP
from time import time, strftime
import subprocess
import getpass
import re, os, sys
import MySQLdb
from colorama import Fore, Style, init
init(autoreset=True) # automatically reset colorama styles etc

# setup some variables
# variables for FTP connection
ftp_host = 'ftp.ensembl.org'
ftp_user = 'anonymous'
ftp_pass = ''
ftp_path = '/pub/current_mysql'

# variables for MySQL connection
sql_host = 'localhost'
sql_port = '3306'
sql_user = 'root'
sql_pass = getpass.getpass(Fore.CYAN + Style.BRIGHT + "\nPlease enter your MySQL password:" + Fore.RESET)

# other variables
data_dir = '/Volumes/DATA/current_mysql'

# start time
start_time = time()

'''
FTP stuff here
'''
print Style.BRIGHT + "\nRetrieving file listings from FTP...\n"

# connect to ftp
ensembl_ftp = FTP(ftp_host) # open connection
ensembl_ftp.login(ftp_user, ftp_pass) # login
ensembl_ftp.cwd(ftp_path) # change to MySQL path

# retrieve directories
ftp_struct = {} # setup dict to store data
dirs = ensembl_ftp.nlst()
for dir in dirs:
	# match only core + ensembl ancestral, compara and ontology dbs if need be
	# remove if and adjust whitespace to retrieve everything
	if re.match("[a-z]+_[a-z]+_core_[0-9]{2}_\w+", dir) or re.match("(^ensembl_[aco][a-z]{6,8}_[0-9]{2}$)", dir):
	# if re.match("^[a-z]+_[a-z]+_[a-z]+_[0-9]{2}_\w+$", dir) or re.match("(^ensembl_[aco][a-z]{6,8}_[0-9]{2}$)", dir):
		print dir
		new_path = os.path.join(ftp_path, dir) # build new path
		ensembl_ftp.cwd(new_path) # change to new path
		files = ensembl_ftp.nlst() # retrieve file listing
		ftp_struct[dir] = files
	
# close ftp connection and quit
ensembl_ftp.quit()

'''
WGET stuff here
'''
print Style.BRIGHT + "\nRetrieving files...\n"
# get files
os.chdir(data_dir) # change to data_dir
keys = ftp_struct.keys() # retrieve all dict keys
keys.sort() # sort alphabetically
for key in keys:
	if not os.path.exists(key):
		os.mkdir(key) # create directory if doesn't exist
	files = ftp_struct[key] # retrieve values (files) from dict
	print key
	for file in files:
		check_file_path = os.path.join(data_dir, key, file[:-3]) # build file path
		if not os.path.exists(check_file_path):
			file_path = os.path.join(data_dir, key) # build file path
			url_path = os.path.join(ftp_path, key, file) # build ftp url
			os.chdir(file_path) # change to file path
			p = subprocess.Popen(["wget", "-t", "0", "-c", "-N", "-nv", "ftp://" + ftp_host + url_path])
			p.wait() # wait for process to finish
	os.chdir(data_dir) # change back to data directory

'''
GUnzip stuff here
'''
print Style.BRIGHT + "\nExtracting files...\n"
os.chdir(data_dir) # change to data_dir
for key in keys:
	print key
	files = ftp_struct[key] # retrieve values (files) from dict
	for file in files:
		check_file_path = os.path.join(data_dir, key, file[:-3]) # build file path
		if not os.path.exists(check_file_path):
			file_path = os.path.join(data_dir, key) # build file path
			os.chdir(file_path)
			p = subprocess.Popen(["gunzip", "-dfq", file])
			p.wait() # wait for process to end
	os.chdir(data_dir)

"""
MySQL import here
"""
print Style.BRIGHT + "\nImporting to MySQL database...\n"
os.chdir(data_dir) # change to data_dir
for key in keys:
	print key
	# create database if not exists
	db = MySQLdb.connect(host=sql_host, port=int(sql_port), user=sql_user, passwd=sql_pass)
	cursor = db.cursor()
	cursor.execute("DROP DATABASE IF EXISTS " + key)
	cursor.execute("CREATE DATABASE IF NOT EXISTS " + key)
	cursor.close()
	# create tables
	sql_file_path = os.path.join(data_dir, key, key + ".sql") # build file path
	os.system("mysql -h " + sql_host + " -u " + sql_user + " -P " + sql_port + " -p" + sql_pass + " " + key + " < " + sql_file_path)
	files = ftp_struct[key] # retrieve values (files) from dict
	for file in files:
		if not file[:-3] == "CHECKSUMS" and not file[:-7] == key:
			file_path = os.path.join(data_dir, key) # build file path
			os.chdir(file_path)
			# import data
			p = subprocess.Popen(["mysqlimport", "-h", sql_host, "-u", sql_user, "-P", sql_port, "--password=" + sql_pass, "--fields_escaped_by=\\\\", key, "-L", file[:-3]])
			p.wait() # wait for process to end
	os.chdir(data_dir)
	
# end time
end_time = time()

# calculate time taken
total_time = end_time - start_time
days, remainder = divmod(total_time, 86400)
hours, remainder = divmod(remainder, 3600)
minutes, seconds = divmod(remainder, 60)
print Fore.YELLOW + Style.BRIGHT + 'Finished in: %d days %d hrs %d mins %d secs' % (days, hours, minutes, seconds)